# RT Anchor — quickstart

Cross-column retention-time calibration for LC-MS lipidomics. Every feature's retention time is mapped onto a **reference column's** time axis (`Cal_RT_min`) and onto a dimensionless 1–100 retention index (`iRT`), so retention becomes comparable across columns, gradients and labs.

This notebook calibrates the **Mix 15** Orbitrap example in [`mix15/`](mix15/) — a sample feature table plus a standards-mixture run from the same column — against the bundled reference column.

## Install

```bash
pip install "rt-anchor[report]"
```

## Calibrate

Two runs from *your* column are required: the sample and the standards mixture, on the same gradient. `panel` names the mixture you ran (`mix15`, `mix21`, or `none` if it is neither); it sets the iRT landmarks and the detection QC, but the calibration itself matches features by m/z alone.

In [1]:
from rt_anchor import calibrate, write_results

res = calibrate(
    "mix15/samples.txt",                    # feature table to calibrate (format auto-detected)
    polarity="positive",
    standards_table="mix15/standards.txt",  # standards run — same column + gradient as the sample
    panel="mix15",                          # mix15 | mix21 | none
)
res.model["method"], res.model["calibration_scope"], res.model["reference"]["label"]

('cross-column-v2', 'project', 'Column 25')

The original table is never altered — calibration only **appends** columns:

In [2]:
cols = ["Cal_RT_min", "Cal_RT_uncertainty_min", "iRT", "iRT_uncertainty",
        "iRT_reliability", "is_extrapolated"]
res.table[[res.mz_col, res.rt_col] + cols].head(8)

,m/z,RT,Cal_RT_min,Cal_RT_uncertainty_min,iRT,iRT_uncertainty,iRT_reliability,is_extrapolated
0,758.5694,7.394,12.474331,0.827495,55.115469,4.063792,low,False
1,765.6162,8.434,13.795925,0.582203,61.605762,2.859172,low,False
2,666.6185,18.037,21.406016,0.325620,98.978598,1.599107,medium,False
3,786.6010,8.762,14.204379,0.590976,63.611662,2.902258,low,False
4,782.5693,7.149,12.182885,0.780965,53.684194,3.835287,low,False
5,496.3394,2.001,4.172998,0.278816,14.347976,1.369254,medium,False
6,759.5718,7.394,12.474331,0.827495,55.115469,4.063792,low,False
7,874.7858,17.791,21.170721,0.183877,97.823075,0.903013,high,False


## What the model did

**Stage 1** matches features between your runs and the reference runs by accurate m/z and fits a robust monotone curve. **Stage 1b — the homologous-series term** — groups the kept pairs into homologous series by Kendrick phase (from m/z alone — these are not identifications), keeps only series that elute in carbon-number order on both columns, and corrects each feature with the curve residuals of the *other* members of its series; it is applied only if a leave-own-out gate shows at least 20% lower error than the curve alone, and on a matrix with few homologous series it simply does not engage. **Stage 2** (endogenous plasma-lipid anchors) is opt-in since 1.2.2 — `use_sample_anchors=True` or the CLI's `--sample-anchors`; when requested, it refines the curve only if leave-one-out error drops by at least 20% (the *gate*). The **iRT** ruler is set by the panel standards found on the reference standards run.

In [3]:
m = res.model
c, s, a, i = m["curve"], m["series"], m["anchors"], m["irt"]
{
    "stage-1 pairs (kept / matched)": f"{c['n_pairs_kept']} / {c['n_pairs']}",
    "series term": s["gate_reason"],
    "validated series (pairs in them / with a leave-own-out prediction)":
        f"{s['n_series']} ({int(res.pairs['in_series'].sum())} / {s['n_pairs_covered']})",
    "features corrected by the series term": s["n_features_corrected"],
    "stage 2 (opt-in)": a["gate_reason"],
    "iRT landmarks on the reference": i["n_landmarks"],
    "landmark span (min)": i["landmark_rt_span_min"],
    "features (extrapolated)": f"{m['n_features']} ({m['n_features_extrapolated']})",
    "iRT reliability": m["confidence_counts"],
}

{'stage-1 pairs (kept / matched)': '433 / 595',
 'series term': 'series term gated off: leave-own-out MSE 35% above curve-only (threshold: 20% below)',
 'validated series (pairs in them / with a leave-own-out prediction)': '27 (182 / 148)',
 'features corrected by the series term': 0,
 'stage 2 (opt-in)': 'anchors disabled',
 'iRT landmarks on the reference': 14,
 'landmark span (min)': [1.455, 21.614],
 'features (extrapolated)': '22314 (569)',
 'iRT reliability': {'medium': 13106, 'low': 6545, 'high': 2663}}

On this example **27 validated series holding 182 of the 433 kept pairs** were found; the gate, read on the 148 of them that get a leave-own-out prediction, **declined** — using them would have raised the leave-own-out error by 35% rather than lowered it, so no feature is corrected and the stage-1 curve is used as-is. That is the expected outcome when homologous series carry no consistent selectivity difference between the two columns — the gate exists for exactly this case. Stage 2 was not requested (`anchors disabled`), so it stayed off.

## Write outputs

`write_results` emits the calibrated table, the fitted model, the curve's evidence (matched pairs, iRT landmarks, stage-2 anchors when requested), a log, and an interactive HTML + static PDF report into [`output/`](output/).

In [4]:
paths = write_results(res, "output/run")
sorted(paths)

['anchors_csv',
 'calibrated_csv',
 'landmarks_csv',
 'log_txt',
 'model_json',
 'pairs_csv',
 'report_html',
 'report_pdf']

| file | content |
|---|---|
| `run_calibrated.csv` | input table + `Cal_RT_min`, `Cal_RT_uncertainty_min`, `iRT`, `iRT_uncertainty`, `iRT_reliability`, `is_extrapolated`, `calibration_scope`, `warp_source`, `RI_spread`, `n_contributing`, `series_correction_min`, `series_n_members` |
| `run_model.json` | curve, the series term's fit and gate decision, stage-2 anchors, iRT definition, reference provenance, detection QC |
| `run_pairs.csv` | every stage-1 matched pair, whether it survived outlier trimming, and whether it sits in a validated homologous series (`in_series`) |
| `run_landmarks.csv` | the panel standards located on the reference standards run (the iRT ruler) |
| `run_anchors.csv` | the stage-2 plasma-lipid anchors and their residuals (empty unless stage 2 was requested) |
| `run_log.txt` | the run log |
| `run_report.html` / `run_report.pdf` | interactive + static report |

See the rendered report at [`output/run_report.html`](output/run_report.html).

## Sanity check — the Mix 21 example

The [`mix21/`](mix21/) example is the bundled reference column's own sample + standards runs. Calibrating a column onto itself must return the input retention time, which makes it a quick end-to-end check of an installation:

In [5]:
ref = calibrate("mix21/samples.txt", polarity="positive",
                standards_table="mix21/standards.txt", panel="mix21")
float((ref.values("Cal_RT_min") - ref.rt_minutes()).abs().max())   # minutes

2.7487345732879476e-11